Sources/NIOCore/ByteBuffer-hexdump.swift

```diff
@@ -0,0 +1,131 @@
+
+extension ByteBuffer {
+
+    /// Return a `String` of hexadecimal digits of bytes in the Buffer,
+    /// in a format that's compatible with `xxd -r -p`.
+    /// `hexDumpShort()` always dumps all readable bytes, i.e. from `readerIndex` to `writerIndex`,
+    /// so you should set those indices to desired location to get the offset and length that you need to dump.
+    func hexDumpShort() -> String {
+        var hexString = ""
+        hexString.reserveCapacity(self.readableBytes*3)
+
+        for byte in self.readableBytesView {
+            hexString += "\(String(byte: byte, padding: 2)) "
+        }
+
+        return String(hexString.dropLast())
+    }
+
+    /// Return a `String` of hexadecimal digits of bytes in the Buffer.,
+    /// in a format that's compatible with `xxd -r -p`, but clips the output to the max length of `limit` bytes.
+    /// If the dump contains more than the `limit` bytes, this function will return the first `limit/2`
+    /// and the last `limit/2` of that, replacing the rest with `...`, i.e. `01 02 03 ... 09 11 12`.
+    ///
+    /// - parameters:
+    ///     - limit: The maximum amount of bytes presented in the dump.
+    func hexDumpShort(limit: Int) -> String {
+        // If the length of the resulting hex dump is smaller than the maximum allowed dump length,
+        // return the full dump. Otherwise, dump the first and last pieces of the buffer, then concatenate them.
+        if readableBytes <= limit {
+            return self.hexDumpShort()
+        } else {
+            let startHex = self.getSlice(at: 0, length: limit/2)!.hexDumpShort()
+            let endHex  = self.getSlice(at: self.readableBytes - limit/2, length: limit/2)!.hexDumpShort()
+            return startHex + " ... " + endHex
+        }
+    }
+
+    /// Returns a `String` of hexadecimal digits of bytes in the Buffer,
+    /// with formatting compatible with output of `hexdump -C`.
+    func hexDumpLong() -> String {
+        var result = ""
+
+        var buffer = self
+        var lineOffset = 0
+
+        while buffer.readableBytes > 0 {
+            let slice = buffer.readSlice(length: min(16, buffer.readableBytes))!
+            let lineLength = slice.readableBytes
+
+            result += "\(String(byte: lineOffset, padding: 8))  "
+
+            // Start with a xxd-format hexdump that we already have, then pad it
+            // and insert a space in the middle.
+            var lineHex = slice.hexDumpShort()
+            lineHex.append(String(repeating: " ", count: 49 - lineHex.count))
+            lineHex.insert(" ", at: lineHex.index(lineHex.startIndex, offsetBy: 24))
+            result += lineHex
+
+            // ASCII column renders the line as ASCII characters, or "." if the character is not printable.
+            result += "|" + String(decoding: slice.getBytes(at: 0, length: slice.readableBytes)!, as: Unicode.UTF8.self).map {
+                $0 >= " " && $0 < "~" ? $0 : "."
```

Comment:
Rather than converting the bytes to string space and then mapping them we can do it more cheaply by operating in byte-space. We can also make use of the readable bytes view:

```swift
let printableBytes = slice.readableBytesView.map { 
    let printableRange = UInt8(ascii: " ") ..< UInt8(ascii: "~")
    return printableRange.contains($0) ? $0 : UInt8(ascii: ".")
}

result += "|" 
result += String(decoding: printableBytes, as: UTF8.self)
result += "|\n"
```

Reply:
I felt greedy with those three lines of string concatenation and squashed them in one line of code. I think readability is still fine, but I appreciate how clear this one by line append code is.

Reply:
IMO readability counts for a lot. I'd rather have five easy to understand lines than a single more complicated line.

Reply:
Oh yes, I absolutely agree with you — that's why I commented on it. I sort of cleaned it up a bit, but still left a single line with string interpolation. But, I'm a Swift newbie, so if you see that three lines are more readable to the folks in the community, I would do that instead.